internal/rbd/group/util.go

```diff
@@ -99,6 +96,106 @@ func (cvg *commonVolumeGroup) initCommonVolumeGroup(
 	cvg.pool = pool
 	cvg.namespace = namespace
 
+	return cvg, nil
+}
+
+// generateVolumeGroupFromMapping checks the clusterID and poolID mapping and
+// generates commonVolumeGroup structure for the mapped clusterID and poolID.
+func generateVolumeGroupFromMapping(
+	ctx context.Context,
+	id string,
+	csiDriver string,
+	creds *util.Credentials,
+	csiID util.CSIIdentifier,
+	mapping *[]util.ClusterMappingInfo,
+) (*commonVolumeGroup, error) {
+	cvg := &commonVolumeGroup{}
+	mcsiID := csiID
+	existingClusterID := csiID.ClusterID
+	existingPoolID := strconv.FormatInt(csiID.LocationID, 10)
+
+	for _, cm := range *mapping {
+		for key, val := range cm.ClusterIDMapping {
+			mappedClusterID := util.GetMappedID(key, val, csiID.ClusterID)
+			if mappedClusterID == "" {
+				continue
+			}
+
+			log.DebugLog(ctx,
+				"found new clusterID mapping %s for existing clusterID %s", mappedClusterID, existingClusterID)
+
+			// Add mapped clusterID to Identifier
+			mcsiID.ClusterID = mappedClusterID
+			for _, pools := range cm.RBDpoolIDMappingInfo {
+				for key, val := range pools {
+					mappedPoolID := util.GetMappedID(key, val, existingPoolID)
+					if mappedPoolID == "" {
+						continue
+					}
+					log.DebugLog(ctx,
+						"found new poolID mapping %s for existing poolID %s", mappedPoolID, existingPoolID)
+
+					mPID, err := strconv.ParseInt(mappedPoolID, 10, 64)
+					if err != nil {
+						return cvg, err
+					}
+					mcsiID.LocationID = mPID
+					cvg, err = generateVolumeGroup(id, csiDriver, creds, mcsiID)
+					if err != nil && !util.ShouldRetryVolumeGeneration(err) {
+						return cvg, err
+					}
+					// If the pool is found, return the volume group
+					if cvg != nil {
+						return cvg, nil
+					}
+				}
+			}
+		}
+	}
+
+	return nil, util.ErrPoolNotFound
+}
+
+func (cvg *commonVolumeGroup) initCommonVolumeGroup(
+	ctx context.Context,
+	id string,
+	csiDriver string,
+	creds *util.Credentials,
+) error {
+	csiID := util.CSIIdentifier{}
+
+	err := csiID.DecomposeCSIID(id)
+	if err != nil {
+		return fmt.Errorf("failed to decompose volume group id %q: %w", id, err)
+	}
+
+	vg, err := generateVolumeGroup(id, csiDriver, creds, csiID)
+	if err != nil && !util.ShouldRetryVolumeGeneration(err) {
+		return err
+	}
+
+	if err != nil && util.ShouldRetryVolumeGeneration(err) {
+		mapping, err := util.GetClusterMappingInfo(csiID.ClusterID)
+		if err != nil {
+			return err
+		}
+		if mapping != nil {
+			vg, err = generateVolumeGroupFromMapping(ctx, id, csiDriver, creds, csiID, mapping)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	cvg.csiDriver = vg.csiDriver
```

Comment:
This is really ugly and rather inefficient. It is also very easy to forget copying an attribute of the struct that was set in an other function.

Can't `generateVolumeGroup()` and `generateVolumeGroup()` not be a function of the `commonVolumeGroup` type so that they update their own instance?